### ISSUE TITLE
W0640 is wrong

### ISSUE THREAD

**Issue by tushar-deepsource** [PyCQA/pylint#7100]
### Bug description

-

### Command used

```shell
pylint asd.py
```

### Pylint output

```shell
asd.py:3:41: W0640: Cell variable num defined in loop (cell-var-from-loop)
```

### Expected behavior

Sounds crazy, but I think `W0640` (`cell-var-from-loop`) is a lint that is non-existent.

Here's the code snippet from the example, which should supposedly cause `i` to be assigned to the last value in the list:

```python
def foo(numbers):
    for i in numbers:
        def bar():
            print(i)
        bar()
```

Except, it works fine:

```pycon
>>> foo([1, 2, 3])
1
2
3
```

Same is true for lambdas:
```python
def foo(numbers):
    for num in numbers:
        print(list(filter(lambda x: x != num, numbers)))
```

```pycon
>>> foo([1, 2, 3])
[2, 3]
[1, 3]
[1, 2]
```

### Pylint version

```shell
pylint 2.13.7
astroid 2.11.3
Python 3.10.4 (main, Apr 26 2022, 19:36:29) [Clang 13.1.6 (clang-1316.0.21.2)]
```

**Comment by tushar-deepsource** [comment]
Documentation aside, it's still a false positive for this simple case:

```python
def foo(numbers):
    for num in numbers:
        print(list(filter(lambda x: x != num, numbers)))
```
Do you think it's safe to not raise this issue for these cases?

**Comment by tushar-deepsource** [comment]
I think I get it now.

```python
def get_functions(numbers):
    for num in numbers:
        function = lambda: print(num)
        yield function

for func in get_functions():
    func()
```

This one works:
```shell
1
2
3
```

This one also works:
```python
def get_functions(numbers):
    for num in numbers:
        function = lambda: print(num)
        yield function

functions = get_functions([1, 2, 3])
for func in functions:
    func()
```
```shell
1
2
3
```

BUT, this one also doesn't work:
```python
def get_functions(numbers):
    for num in numbers:
        function = lambda: print(num)
        yield function

functions = list(get_functions([1, 2, 3]))
for func in functions:
    func()
```
```shell
3
3
3
```

It's because `list(...)` eagerly collects the functions, but for-loops lazily yield the functions. If you eagerly evaluate the functions, the value of `num` gets to `3` before the function is ever called.

**Comment by Pierre-Sassoulas** [comment]
https://github.com/PyCQA/pylint/issues/5263 is a near duplicate.

**Comment by mthuurne** [comment]
Note that you can avoid this issue by capturing the variable's value as an argument's default value:
```py
def foo(numbers):
    for num in numbers:
        print(list(filter(lambda x, n=num: x != n, numbers)))
```
As the lambda no longer uses variables from the context in which it was created (cell vars), there is no risk of the variable's value changing underneath it, no matter whether the lambda is called within the current loop iteration or later.
